[WIP] remove emoji from changelog

koenpunt opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

When trying to submit a changelog containing emoji, the request fails with a response similar to:

```
ERROR-1: What to Test can't contain the following characters: ?.; ERROR-1: What to Test can't contain the following characters: ?.; ERROR-1: What to Test can't contain the following characters: ?.
```

### Description

Apple doesn't allow emoji to be used in the changelog value

WIP because the regex has to be updated to apply to a broader range of emoji, or maybe there's a gem that can do the heavy lifting for us.

getaaron reviewed (changes requested):
I don't think we should arbitrarily change the changelog content. How about something like:

```ruby
UI.user_error!("Disappointingly, emojis aren't allowed in the changelog 🙄") if options[:changelog].match(/[\u{1F600}-\u{1F6FF}]/)
```

joshdholtz commented:
@koenpunt I agree with @getaaron on this one - we tend to stay away from automatically changing things on the use. **BUT**, since you said that this was from a changeling generated from git commands, it maybe be best to add a cleanse/sanitize emojis option in the `changelog_from_git_commits` action?

koenpunt commented:
Since emoji cause the problem for Apple, the flag should be with the `set_changelog` and `upload_to_testflight` actions IMO.
But since we're already truncating the changelog based on size, I don't really see a problem with just replacing the emoji, especially since it otherwise doesn't work. 
We could of course still warn the user about the fact that those symbols have been removed, which is what is what also happens when the changelog is truncated.

joshdholtz commented:
@koenpunt That is a fair point ^ I think if you can check if there is an emoji and then warn that user (with your current implementation) that be 👌

koenpunt committed:
remove Emoji symbols from changelog

Emoji are not allowed by Apple, and submitting a changelog that contains them will result in an error. When the changelog now contains Emoji, they will be removed automatically, and a message is shown to the user that the removal has taken place.

koenpunt commented:
Alright, I've updated the implementation

getaaron commented:
We might want to add a unit test to [build_manager_spec.rb](https://github.com/fastlane/fastlane/blob/6649d0daa6d8c51105c33d380aa526b657f12d6f/pilot/spec/build_manager_spec.rb)?

e.g.

```
  it "removes emoji" do
    changelog = "I'm 🦇B🏧an!"
    changelog = Pilot::BuildManager.truncate_changelog(changelog)
    expect(changelog).to eq("I'm Ban!")
  end

  it "truncates spaces after removing emoji at the end" do
    changelog = "Why doesn't Apple support Emoji? 🤷‍♀️"
    changelog = Pilot::BuildManager.truncate_changelog(changelog)
    expect(changelog).to eq("Why doesn't Apple support Emoji?")
  end
```

…and now that I think about it, maybe `truncate_changelog` should come after the emojis are stripped?

joshdholtz reviewed (dismissed):
Yeah yeah, I agree with @getaaron. Can we add some tests to this and call truncate after sanitize?

joshdholtz commented:
@koenpunt Heyyy 👋 Are you still working on this PR? 😊

koenpunt committed:
change order of emoji stripping and truncating

koenpunt committed:
add tests for sanitize_changelog

getaaron commented:
Thanks, @koenpunt. Great work!

koenpunt commented:
> Thanks, @koenpunt. Great work!

You're welcome! 🎉
